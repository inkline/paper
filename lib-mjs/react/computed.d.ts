import { Ref } from '@inkline/paper/types';
export declare function computed<T>(computeFn: () => T): Ref<T>;
