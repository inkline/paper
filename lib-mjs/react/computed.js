export function computed(computeFn) {
    return {
        get value() {
            return computeFn();
        }
    };
}
//# sourceMappingURL=computed.js.map