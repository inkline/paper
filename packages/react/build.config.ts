export default {
    entries: [
        {
            builder: 'mkdist',
            format: 'esm',
            ext: 'mjs',
            input: './src',
            declaration: true
        },
        {
            builder: 'mkdist',
            format: 'cjs',
            ext: 'cjs',
            input: './src',
            declaration: true
        }
    ]
};
