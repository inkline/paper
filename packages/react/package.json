{
  "name": "@inkline/paper-react",
  "description": "Inkline Paper React.js package.",
  "scripts": {
    "pre": "cp ../common/src/types.ts src/types.common.ts",
    "pretest": "npm run pre",
    "test": "vitest run",
    "prebuild": "npm run pre",
    "build": "tsc -d -p tsconfig.mjs.json && tsc -d -p tsconfig.cjs.json && mjscjs -p tsconfig.mjs.json && mjscjs -p tsconfig.cjs.json"
  },
  "devDependencies": {
    "@testing-library/react": "12.1.3",
    "@vitejs/plugin-react": "1.2.0",
    "jsdom": "19.0.0",
    "mjscjs": "1.0.9",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "rimraf": "3.0.2",
    "vite": "2.8.6",
    "vitest": "0.5.9"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/inkline/paper.git"
  },
  "license": "ISC"
}
