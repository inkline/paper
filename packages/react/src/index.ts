export * from './computed';
export * from './defineComponent';
export * from './definePlugin';
export * from './events';
export * from './fragment';
export * from './h';
export * from './helpers';
export * from './provide';
export * from './ref';
export * from './types';
export * from './watch';
