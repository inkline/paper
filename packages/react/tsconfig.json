{
    "extends": "../../tsconfig.json",
    "include": [
        "src/**/*.ts",
        "src/**/*.tsx"
    ],
    "compilerOptions": {
        "rootDir": "src",
        "baseUrl": ".",
        "jsx": "preserve",
        "jsxFactory": "h",
        "jsxFragmentFactory": "Fragment",
        "types": [
            "@types/react"
        ],
        "paths": {
            "@inkline/paper": ["@inkline/paper/src"]
        }
    },
    "references": [
        {
            "path": "./tsconfig.vite.json"
        }
    ]
}
