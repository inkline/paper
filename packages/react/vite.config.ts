import { defineConfig } from 'vite';
import { resolve } from 'path';
import react from '@vitejs/plugin-react';

// https://vitejs.dev/config/
export default defineConfig({
    plugins: [
        react()
    ],
    resolve: {
        alias: [
            {
                find: /^@inkline\/paper\//,
                replacement: `${resolve(__dirname)}/src/`
            }
        ]
    },
    build: {
        lib: {
            entry: resolve(__dirname, 'src', 'index.ts'),
            name: 'InklinePaperReact',
            fileName: (format) => `react.${format}.js`
        },
        rollupOptions: {
            // make sure to externalize deps that shouldn't be bundled
            // into your library
            external: [
                'react'
            ],
            output: {
                exports: 'named',
                // Provide global variables to use in the UMD build
                // for externalized deps
                globals: {
                    react: 'React'
                }
            }
        }
    },
    test: {
        globals: true,
        environment: 'jsdom'
    }
});
