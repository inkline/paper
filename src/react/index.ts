export * from '@inkline/ucd/react/defineComponent';
export * from '@inkline/ucd/react/computed';
export * from '@inkline/ucd/react/h';
export * from '@inkline/ucd/react/helpers';
export * from '@inkline/ucd/react/ref';
export * from '@inkline/ucd/react/types';
