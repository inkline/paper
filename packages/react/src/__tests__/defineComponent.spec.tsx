import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import {
    defineComponent,
    h,
    Fragment,
    ref,
    Events
} from '../index';
import { fireEvent, render } from '@testing-library/react';

describe('react', () => {
    describe('defineComponent()', () => {
        it('should render component', () => {
            const Component = defineComponent({
                render () {
                    return <div />;
                }
            });

            const wrapper = render(<Component />);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        it('should render component with state', () => {
            const Component = defineComponent({
                setup () {
                    const count = ref(0);

                    return { count };
                },
                render (state) {
                    return <button>{state.count.value}</button>;
                }
            });

            const wrapper = render(<Component />);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        it('should render component with state and props', () => {
            const Component = defineComponent({
                setup (props) {
                    const count = ref(props.initialValue);

                    return { count };
                },
                render (state) {
                    return <button>{state.count.value}</button>;
                }
            });

            const wrapper = render(<Component initialValue={1} />);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        it('should render component with state, props, and onClick state function', async () => {
            const Component = defineComponent({
                setup (props) {
                    const count = ref(props.initialValue);

                    const onClick = () => {
                        count.value += 1;
                    };

                    return { count, onClick };
                },
                render (state) {
                    return <button onClick={state.onClick}>{state.count.value}</button>;
                }
            });

            const wrapper = render(<Component initialValue={1} />);
            await fireEvent.click(wrapper.container.firstChild as Element);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        describe('context', () => {
            describe('slot()', () => {
                it('should render react component with default slot, one child', () => {
                    const Component = defineComponent({
                        render (state, ctx) {
                            return <div>
                                {ctx.slot()}
                            </div>;
                        }
                    });

                    const wrapper = render(<Component>
                        <span>Child</span>
                    </Component>);
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });

                it('should render react component with default slot, multiple children', () => {
                    const Component = defineComponent({
                        render (state, ctx) {
                            return <div>
                                {ctx.slot()}
                            </div>;
                        }
                    });

                    const wrapper = render(<Component>
                        <span>Child 1</span>
                        <span>Child 2</span>
                        <span>Child 3</span>
                    </Component>);
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });

                it('should render react component with named default slot', () => {
                    const Component = defineComponent({
                        render (state, ctx) {
                            return <div>
                                {ctx.slot()}
                            </div>;
                        }
                    });

                    const wrapper = render(<Component>{{
                        default: () => <span>Child</span>
                    }}</Component>);
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });

                it('should render react component with named slots', () => {
                    const Component = defineComponent({
                        slots: [
                            'header',
                            'footer'
                        ],
                        render (state, ctx) {
                            return <div>
                                {ctx.slot('header')}
                                {ctx.slot()}
                                {ctx.slot('footer')}
                            </div>;
                        }
                    });

                    const wrapper = render(<Component>{{
                        header: () => <span>Header</span>,
                        default: () => <span>Body</span>,
                        footer: () => <span>Footer</span>
                    }}</Component>);
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });
            });

            describe('emit()', () => {
                it('should emit event by attaching on[EventName] callback', async () => {
                    const Component = defineComponent({
                        emits: [
                            'click'
                        ],
                        setup (props, ctx) {
                            const onClick = () => {
                                ctx.emit('click');
                            };

                            return { onClick };
                        },
                        render (state) {
                            return <button onClick={state.onClick}>Button</button>;
                        }
                    });

                    const onClick = vi.fn();
                    const wrapper = render(<Component onClick={onClick} />);
                    await fireEvent.click(wrapper.container.firstChild as Element);
                    expect(onClick).toHaveBeenCalled();
                });

                it('should emit event with arguments', async () => {
                    const Component = defineComponent({
                        emits: [
                            'click'
                        ],
                        setup (props, ctx) {
                            const onClick = (event: Event) => {
                                ctx.emit('click', event);
                            };

                            return { onClick };
                        },
                        render (state) {
                            return <button onClick={state.onClick}>Button</button>;
                        }
                    });

                    const onClick = vi.fn();
                    const wrapper = render(<Component onClick={onClick} />);
                    await fireEvent.click(wrapper.container.firstChild as Element);
                    expect(onClick).toHaveBeenCalled();
                    expect(onClick.mock.calls[0][0]).toEqual(expect.any(Object));
                });

                it('should render component with modelValue and update:modelValue', async () => {
                    const Component = defineComponent({
                        emits: [
                            'update:modelValue'
                        ],
                        setup (props, ctx) {
                            const onClick = () => {
                                ctx.emit('update:modelValue', props.modelValue + 1);
                            };

                            return { onClick };
                        },
                        render (state) {
                            return <button onClick={state.onClick}>Button</button>;
                        }
                    });

                    let value = 3;
                    const onUpdate = vi.fn((newValue) => { value = newValue; });
                    const wrapper = render(<Component modelValue={value} onUpdateModelValue={onUpdate} />);

                    await fireEvent.click(wrapper.container.firstChild as Element);

                    expect(onUpdate).toHaveBeenCalled();
                    expect(value).toEqual(4);
                });

                it('should render component with input field', async () => {
                    const Component = defineComponent({
                        props: {
                            modelValue: {
                                type: String,
                                default: ''
                            }
                        },
                        emits: [
                            'update:modelValue'
                        ],
                        setup (props, ctx) {
                            const onChange = (event: Event) => {
                                ctx.emit('update:modelValue', (event.target as HTMLInputElement).value);
                            };

                            return { onChange };
                        },
                        render (state) {
                            const events = {
                                [Events.onInputChange]: state.onChange
                            };

                            return <input value={state.modelValue} {...events} />;
                        }
                    });

                    let value = '';
                    const onUpdate = vi.fn((newValue) => { value = newValue; });
                    const wrapper = render(<Component modelValue={value} onUpdateModelValue={onUpdate} />);

                    await fireEvent.change(wrapper.container.firstChild as Element, { target: { value: 'abc' } });

                    expect(onUpdate).toHaveBeenCalled();
                    expect(value).toEqual('abc');
                });
            });
        });

        describe('provide/inject()', () => {
            it('should provide data to children', async () => {
                const identifier = Symbol('provide-reactive');
                const Provider = defineComponent({
                    setup (props, ctx) {
                        ctx.provide(identifier, 'value', []);

                        return {};
                    },
                    render (state, ctx) {
                        return <div>{ctx.slot()}</div>;
                    }
                });

                const Consumer = defineComponent({
                    setup (props, ctx) {
                        const providedValue = ctx.inject(identifier);

                        return { providedValue };
                    },
                    render (state) {
                        return <div>{state.providedValue}</div>;
                    }
                });

                const wrapper = render(<Provider>
                    <Consumer key={'consumer'} />
                </Provider>);
                expect(wrapper.container.firstChild).toMatchSnapshot();
            });

            it('should provide reactive data to children', async () => {
                const identifier = Symbol('provide-reactive');
                const Provider = defineComponent({
                    setup (props, ctx) {
                        const count = ref(0);
                        const onClick = () => { count.value += 1; };

                        ctx.provide(identifier, count, [count.value]);

                        return { onClick };
                    },
                    render (state, ctx) {
                        return <button onClick={state.onClick}>{ctx.slot()}</button>;
                    }
                });

                const Consumer = defineComponent({
                    setup (props, ctx) {
                        const providedValue = ctx.inject(identifier);

                        return { providedValue };
                    },
                    render (state) {
                        return <div>{state.providedValue?.value}</div>;
                    }
                });

                const wrapper = render(<Provider>
                    <Consumer key={'consumer'} />
                </Provider>);
                await fireEvent.click(wrapper.container.firstChild as Element);
                expect(await wrapper.findByText('1')).toBeTruthy();
                expect(wrapper.container.firstChild).toMatchSnapshot();
            });

            it('should overwrite reactive data for deeply nested children', async () => {
                const identifier = Symbol('provide-reactive');
                const ProviderA = defineComponent({
                    setup (props, ctx) {
                        const count = ref(0);

                        ctx.provide(identifier, count, [count.value]);

                        return {};
                    },
                    render (state, ctx) {
                        return <div>{ctx.slot()}</div>;
                    }
                });

                const ProviderB = defineComponent({
                    setup (props, ctx) {
                        const count = ref(1);
                        const onClick = () => { count.value += 1; };

                        ctx.provide(identifier, count, [count.value]);

                        return { onClick };
                    },
                    render (state, ctx) {
                        return <button onClick={state.onClick}>
                            {ctx.slot()}
                        </button>;
                    }
                });

                const Consumer = defineComponent({
                    setup (props, ctx) {
                        const providedValue = ctx.inject(identifier);

                        return { providedValue };
                    },
                    render (state) {
                        return <div>{state.providedValue?.value}</div>;
                    }
                });

                const wrapper = render(<ProviderA>
                    <ProviderB>
                        <Consumer key={'consumer'} />
                    </ProviderB>
                </ProviderA>);
                await fireEvent.click(wrapper.container.querySelector('button') as Element);
                expect(await wrapper.findByText('2')).toBeTruthy();
                expect(wrapper.container.firstChild).toMatchSnapshot();
            });

            it('should provide reactive data based on id', async () => {
                const Provider = defineComponent({
                    setup (props, ctx) {
                        const text = ref(props.id);
                        const onClick = () => { text.value = 'abc'; };

                        ctx.provide(props.id, text, [text.value]);

                        return { onClick };
                    },
                    render (state, ctx) {
                        return <button onClick={state.onClick}>{ctx.slot()}</button>;
                    }
                });

                const Consumer = defineComponent({
                    setup (props, ctx) {
                        const providedValue = ctx.inject(props.id);

                        return { providedValue };
                    },
                    render (state) {
                        return <div>{state.providedValue?.value}</div>;
                    }
                });

                const wrapper = render(<div>
                    <Provider id={'a'}>
                        <Consumer id={'a'} />
                    </Provider>
                    <Provider id={'b'}>
                        <Consumer id={'b'}/>
                    </Provider>
                </div>);

                const buttons = wrapper.container.querySelectorAll('button');
                expect(wrapper.container.firstChild).toMatchSnapshot();
                await fireEvent.click(buttons[0] as Element);
                expect(wrapper.container.firstChild).toMatchSnapshot();
                await fireEvent.click(buttons[1] as Element);
                expect(wrapper.container.firstChild).toMatchSnapshot();
            });
        });
    });
});
