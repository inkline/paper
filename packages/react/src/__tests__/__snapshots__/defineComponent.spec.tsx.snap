// Vitest Snapshot v1

exports[`react > defineComponent() > context > slot() > should render react component with default slot, multiple children 1`] = `
<div>
  <span>
    Child 1
  </span>
  <span>
    Child 2
  </span>
  <span>
    Child 3
  </span>
</div>
`;

exports[`react > defineComponent() > context > slot() > should render react component with default slot, one child 1`] = `
<div>
  <span>
    Child
  </span>
</div>
`;

exports[`react > defineComponent() > context > slot() > should render react component with named default slot 1`] = `
<div>
  <span>
    Child
  </span>
</div>
`;

exports[`react > defineComponent() > context > slot() > should render react component with named slots 1`] = `
<div>
  <span>
    Header
  </span>
  <span>
    Body
  </span>
  <span>
    Footer
  </span>
</div>
`;

exports[`react > defineComponent() > provide/inject() > should overwrite reactive data for deeply nested children 1`] = `
<div>
  <button>
    <div>
      2
    </div>
  </button>
</div>
`;

exports[`react > defineComponent() > provide/inject() > should provide data to children 1`] = `
<div>
  <div>
    value
  </div>
</div>
`;

exports[`react > defineComponent() > provide/inject() > should provide reactive data based on id 1`] = `
<div>
  <button>
    <div>
      a
    </div>
  </button>
  <button>
    <div>
      b
    </div>
  </button>
</div>
`;

exports[`react > defineComponent() > provide/inject() > should provide reactive data based on id 2`] = `
<div>
  <button>
    <div>
      abc
    </div>
  </button>
  <button>
    <div>
      b
    </div>
  </button>
</div>
`;

exports[`react > defineComponent() > provide/inject() > should provide reactive data based on id 3`] = `
<div>
  <button>
    <div>
      abc
    </div>
  </button>
  <button>
    <div>
      abc
    </div>
  </button>
</div>
`;

exports[`react > defineComponent() > provide/inject() > should provide reactive data to children 1`] = `
<button>
  <div>
    1
  </div>
</button>
`;

exports[`react > defineComponent() > should render component 1`] = `<div />`;

exports[`react > defineComponent() > should render component with state 1`] = `
<button>
  0
</button>
`;

exports[`react > defineComponent() > should render component with state and props 1`] = `
<button>
  1
</button>
`;

exports[`react > defineComponent() > should render component with state, props, and onClick state function 1`] = `
<button>
  2
</button>
`;
